OpenRA.Mods.Common/Traits/FireWarheadOnDeath.cs

```diff
@@ -132,6 +140,13 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
 			}
 		}
 
+		Actor GetDamageSource(Actor self, AttackInfo e)
+		{
+			return Info.DamageSource == DamageSource.Killer ? e.Attacker
+				: Info.DamageSource == DamageSource.Self || parent == null || parent.IsDead ? self
+				: parent;
```

Comment:
This seems correct, but it would be a lot less confusing to read with a switch statement based on `DamageSource` and then not mixing cases.